ext/snmp/snmp.c

```diff
@@ -936,7 +939,35 @@ static bool netsnmp_session_set_auth_protocol(struct snmp_session *s, zend_strin
 		return true;
 	}
 
-	zend_value_error("Authentication protocol must be either \"MD5\" or \"SHA\"");
+#ifdef HAVE_SNMP_SHA256
+	if (zend_string_equals_literal_ci(prot, "SHA256")) {
+		s->securityAuthProto = usmHMAC192SHA256AuthProtocol;
+		s->securityAuthProtoLen = sizeof(usmHMAC192SHA256AuthProtocol) / sizeof(oid);
+		return true;
+	}
+#endif
+
+#ifdef HAVE_SNMP_SHA512
+	if (zend_string_equals_literal_ci(prot, "SHA512")) {
+		s->securityAuthProto = usmHMAC384SHA512AuthProtocol;
+		s->securityAuthProtoLen = sizeof(usmHMAC384SHA512AuthProtocol) / sizeof(oid);
+		return true;
+	}
+#endif
+
+	smart_string_appendl(&err, "Authentication protocol must be either \"SHA\"", 44);
+#ifdef HAVE_SNMP_SHA256
+	smart_string_appendl(&err, " or \"SHA256\"", 12);
+#endif
+#ifdef HAVE_SNMP_SHA512
+	smart_string_appendl(&err, " or \"SHA512\"", 12);
+#endif
+#ifndef DISABLE_MD5
+	smart_string_appendl(&err, " or \"MD5\"", 9);
+#endif
+	smart_string_0(&err);
+	zend_value_error("%s", err.c);
+	smart_string_free(&err);
```

Comment:
Can we separate this so that if there is an unrecognized protocol (like md56, which I accidentally typo'd), that it is a separate error message from passing SHA512 but not having SHA512 support on this particular build?
A small grammar nit: either-or is used for exactly 2 options; simply remove the word "either".
Looks good to me otherwise.

Reply:
> separate error message from passing SHA512 but not having SHA512 support on this particular build?

Perhaps more complexity for low benefit....

> simply remove the word "either".

Done (and squashed)